### ISSUE TITLE
unw_backtrace() returns wrong addresses

### ISSUE THREAD

**Issue by perfhunter** [libunwind/libunwind#94]
I have unw_backtrace() on x86-64 machine always returning addresses that are off by one relative to the instruction address after the call. This issue is related to fast backtrace optimization and I guess it may be caused by not restoring rip after decrementing it by use_prev_instr flag. Indeed, in src/x86_64/Gtrace.c we have:
   line 438:   rip -= d->use_prev_instr;
and before finishing the loop iteration:
   line 543:   buffer[depth++] = (void *) (rip - d->use_prev_instr);
I see no place where rip gets incremented again in this loop, so at least for standard frames we definitely write to the buffer the wrong value.
There is no problem if I manually unwind the stack using unw_step().

**Comment by djwatson** [comment]
We've had some recent fixes around this, does this issue still exist in master?  Thanks

**Comment by tyroguru** [comment]
I've just ran into what I think is this problem - all the stack pointers returned are off-by-one byte. At the minute I can't be sure what version of libunwind we are linking against but I expect it to be fairly recent (this is Facebook's internal jemalloc build). Obviously it's just an minor annoyance more than anything but I'd be interested in if there were any thoughts on a fix.

**Comment by djwatson** [comment]
IIRC I ran in to this once at facebook - the root cause was bad .eh_frame annotations in glibc assembly files.  Or something like that. I don't think it ever got fixed. Ask kenny yu

**Comment by tyroguru** [comment]
Thanks Dave! I'll speak to Kenny.

**Comment by kennyyu** [comment]
Hey there! For the investigation at fb, we were finding issues in unwinding `__lll_lock_wait()` calls in glibc, and we discovered that the frame pointers were still correct, but the debug information was broken. If you have a current stack you want to symbolize, using the workaround here might work: https://stackoverflow.com/questions/42739893/force-gdb-to-use-frame-pointer-based-unwinding/42741367

@djwatson Sad you left FB! Hope you're doing well! For the `__lll_lock_wait()` fix, I can try to submit my glibc patches upstream to fix the `__lll_lock_wait()` issue -- would you be able to help answer questions when I have them?

**Comment by tyroguru** [comment]
I think the problem this issue describes is a seperate one to the one Kenny describes. This is specifically that the addresses returned by the unwinder are always 1 byte off what they should actually be. If I apply a different unwinder to the target process such as bpf (via bpftrace) then the stacks are all good. It certainly seems to be a libunwinder issue.

**Comment by djwatson** [comment]
> If I apply a different unwinder to the target process such as bpf (via bpftrace) then the stacks are all good. It certainly seems to be a libunwinder issue.

Are you sure you're still using an dwarf based unwinder? with perf this requires --call-graph=dwarf, and is like 10x as slow since it requires dumping the stack memory to userspace (the kernel doesn't have a dwarf unwinder like it does an fp unwinder).

The issue described above only affects dwarf unwinders, not fp.

**Comment by tyroguru** [comment]
Dave - I don't know this code or libunwind at all so thanks for your help! I'm stepping through our stack walking and the addresses are definitely obtained through x86_64/tdep_trace(). It's definite that the rip's being stored are off-by-one because of this line x86_64/Gtrace.c:543:

buffer[depth++] = (void *) (rip - d->use_prev_instr);

Should we be using a DWARF based unwinder as I would have expected us to just be using a fp unwinder. Is our libunwinder built incorrectly? Again, apologies for my lack of knowledge here.

**Comment by djwatson** [comment]
> It's definite that the rip's being stored are off-by-one because of this line x86_64/Gtrace.c:543:
> buffer[depth++] = (void *) (rip - d->use_prev_instr);

use_prev_instr is set when libunwind thinks this is a signal frame.  Without a repro I don't know why this is, but I'm suggesting you might have incorrect dwarf info from hand-coded assembly.

> Should we be using a DWARF based unwinder as I would have expected us to just be using a fp unwinder. Is our libunwinder built incorrectly? Again, apologies for my lack of knowledge here.

I don't have a horse in this race, use whatever unwinder you want.   No, I don't think your libunwind is built incorrectly.

**Comment by djwatson** [comment]
One other thought - you could try the current github HEAD, there's been lots of fixes - I've been meaning to roll a new release sometime

**Comment by tyroguru** [comment]
Thanks again. I'll speak to the team and see what we can do about getting the latest bits and testing.

**Comment by jasonbking** [comment]
I'm seeing the exact same issue using libunwind on illumos as well -- running the Gtest-trace test in verbose mode, the 'normal trace' values are all 1 byte more than the values from either backtrace() or unw_backtrace() for all three tests.

There is no hand coded assembly in any of the stack frames in the test on illumos either -- all the eh_frame data was created by gcc7.

**Comment by jasonbking** [comment]
Also, disabling the weak backtrace symbol -- and using the illumos libc backtrace() implementation (which strictly uses frame pointers), the 'normal trace' and backtrace() addresses agree, while the unw_backtrace() addresses are one byte less.

**Comment by jasonbking** [comment]
Digging a bit more.. it looks like the signal frame detection for illumos and Solaris isn't correct, so that's probably at least part of the issue there..

**Comment by tyroguru** [comment]
@jasonbking - what are you basing your signal frame detection comment on? I'm obviously not connecting the dots properly hereso would be very interested in your debugging observations. Thanks.

**Comment by jasonbking** [comment]
Based on the illumos source -- basically the kernel pushes an invalid IP on the stack, so when walking the stack, you'll get an IP of -1 when you hit a signal frame.

I've filed a PR with fixes for the ones I've found so far.

**Comment by tyroguru** [comment]
I've just spent a bit more time looking at this and I just can't get away from my observation (and the initial observation in this issue made by @perfhunter) that it is incorrect to adjust the rip value being placed into the buffer array.

 'use_prev_instr' is required so that trace_lookup() and descendants can correctly identify the boundary of the function that the current rip sits in, adjusting it for the type of frame (regular or signal really). In x86_64/tdep_trace(), when the next frame has been correctly looked up we extract the rip according to the frame type and set 'use_prev_instr' for the next go around. However, the rip that is inserted into the 'buffer' array never needs adjusting as it is the *actual* rip value. As I said, 'use_prev_instr' is simply to control the range of instructions being considered when looking up FDEs and shouldn't be applied to the actual stored address that we pulled out of the frame.

I think the fix is as simple as removing the adjustment to the stored rip value:

```
diff --git a/src/x86_64/Gtrace.c b/src/x86_64/Gtrace.c
index 74122710..10177395 100644
--- a/src/x86_64/Gtrace.c
+++ b/src/x86_64/Gtrace.c
@@ -540,7 +540,7 @@ tdep_trace (unw_cursor_t *cursor, void **buffer, int *size)
       break;

     /* Record this address in stack trace. We skipped the first address. */
-    buffer[depth++] = (void *) (rip - d->use_prev_instr);
+    buffer[depth++] = (void *)rip;
   }

 #if UNW_DEBUG
```

@djwatson - we thought that this may have been a FB specific thing but I've tested my test cases on a non-FB system (Fedora) and the problem manifests in the same manner.

**Comment by djwatson** [comment]
I am convinced.  Put up a diff.

./configure --disable-weak-backtrace
make check -j4
./test/Gtest-trace -v

Indeed the unw_step, backtrace, and unw_backtrace results disagree.

Seems to blame to 5f38f35d, but that was a long time ago?

**Comment by tyroguru** [comment]
Thanks @djwatson for the pointer to the test - better than using my own hand cranked tests!

I just created the pull request. Yes, it would appear that 5f38f35 introduced this and that this problem has existed for 9 years! I guess people don't often cut and paste stack addresses into gdb and actually look at the instructions given back to them at that address :-) .

**Comment by djwatson** [comment]
Thanks everyone for digging and fixing this, and especially @tyroguru for the pull request.  Merged, and will put it in 1.5-rc2 as well.
